### PR TITLE
C99 support

### PR DESCRIPTION
**Pull request by nwc10**
Tested on *nix.

Needs testing on Win32 and VMS. Also needs the Win32 documentation changes reviewed by someone familiar with Win32

### CODE REVIEW

File: README.win32
Comment (steve-m-hay) [inline README.win32]: The "or "is no longer wanted here.

File: README.win32
Comment (steve-m-hay) [inline README.win32]: MinGW-w64 _is_ still being tested.

File: README.win32
Comment (xenu) [inline README.win32]: Visual Studio 2013 is the only version that still has Express edition (in addition to Community), so mentioning both seems overly prudent. I think "Express/Community" can be now simplified to "Community" everywhere in this document.

**Review by smpetershm — COMMENTED** [review]
This makes me twitch after adding that warning so many years ago.

**Review by steve-m-hay — CHANGES_REQUESTED** [review]
win32/GNUmakefile and win32/Makefile also have large chunks that can be removed.
Likewise many .c and .h files (not just in win32/). VC12 has _MSC_VER 1800...

File: README.win32
Comment (nwc10) [reply README.win32]: The full context was:

```
This port can also be built on IA64/AMD64 using:

      Microsoft Platform SDK    Nov 2001 (64-bit compiler and tools)
      MinGW64 compiler (gcc version 4.4.3 or later)
```

and that text (as-was) is already ambiguous, isn't it? In that **IA64** is Itanium (and long dead for Windows) whereas **AMD64** is the new normal. So is it fair to say that we should just take out the whole thing, and change the earlier text:

```
This set of instructions is meant to describe a so-called "native"
port of Perl to the Windows platform.  This includes both 32-bit and
64-bit Windows operating systems.  The resulting Perl requires no
additional software to run (other than what came with your operating
system).  Currently, this port is capable of using one of the
following compilers on the Intel x86 architecture:

      Microsoft Visual C++    version 12.0 or later
      Intel C++ Compiler      (experimental)
      Gcc by mingw.org        gcc version 3.4.5-5.3.0
      Gcc by mingw-w64.org    gcc version 4.4.3 or later

Note that the last two of these are actually competing projects both
delivering complete gcc toolchain for MS Windows:
```

to reflect both on x86 and x86_64?


Or should instead keep that as-is and change the first text to:

```
This port can also be built on AMD64 using:

      MinGW64 compiler (gcc version 4.4.3 or later)
```

just with nicer formatting?

**Comment by nwc10** [comment]
> This makes me twitch after adding that warning so many years ago.

Which particular warning? github didn't hook your comment up to any particular line of change, and I caused so much code churn that I'm failing to figure out which it was.

And i this twitch a good thing, or a bad thing? (Should I be ready to hide?)

**Comment by nwc10** [comment]
> win32/GNUmakefile and win32/Makefile also have large chunks that can be removed. Likewise many .c and .h files (not just in win32/). VC12 has _MSC_VER 1800...

Oh my. That's a good thing, isn't it?

However, I'm not competent to do that. Is it better if we wait to do that in a follow up branch/PR?

File: README.win32
Comment (xenu) [reply README.win32]: The former sounds good to me. There's nothing special about AMD64.

**Comment by jkeenan** [comment]
Today I compiled perl on the `smoke-me/nicholas/C99` branch in order to see what effect merging this branch would have on build-time warnings.

Good news:  I tried this branch on FreeBSD-12 using both `clang10` (the default C compiler) and `clang12`.  There were no changes in build-time warnings emitted compared with blead.

Almost as good news:  I tried this branch on FreeBSD-12 using `gcc9` (the default version of `gcc` on this platform).  There were 3 new instances of the `-Wdeclaration-after-statement` warnings category.  I suspect these warnings can be remediated fairly easily.

**blead | Sep 17 2021 | freebsd | gcc9**
```
$ report-build-warnings 47e2d53c92.freebsd.gcc9.threaded.maketp.output.txt.gz
File:  47e2d53c92.freebsd.gcc9.threaded.maketp.output.txt.gz

  Wc++-compat                              240
  Wempty-body                                6
  Wsign-compare                              4
  Wtype-limits                               2

```
**smoke-me/nicholas/C99 | Oct 06 2021 | freebsd | gcc9**
```
$ report-build-warnings 6b5b56c60c.freebsd.gcc9.threaded.smoke-me-nicholas-C99.maketp.output.txt.gz
File:  6b5b56c60c.freebsd.gcc9.threaded.smoke-me-nicholas-C99.maketp.output.txt.gz

  Wc++-compat                              240
  Wdeclaration-after-statement               3
  Wempty-body                                6
  Wsign-compare                              4
  Wtype-limits                               2

$ ~/bin/perl/parse-one-warnings-group 6b5b56c60c.freebsd.gcc9.threaded.smoke-me-nicholas-C99.maketp.output.txt.gz Wdeclaration-after-statement
File:  6b5b56c60c.freebsd.gcc9.threaded.smoke-me-nicholas-C99.maketp.output.txt.gz

[
  {
    char   => 5,
    group  => "Wdeclaration-after-statement",
    line   => 91,
    source => "../../inline.h",
    text   => "ISO C90 forbids mixed declarations and code",
  },
  {
    char   => 5,
    group  => "Wdeclaration-after-statement",
    line   => 91,
    source => "../../inline.h",
    text   => "ISO C90 forbids mixed declarations and code",
  },
  {
    char   => 5,
    group  => "Wdeclaration-after-statement",
    line   => 91,
    source => "../../inline.h",
    text   => "ISO C90 forbids mixed declarations and code",
  },
]
```
These 3 new warnings are being emitted from compilation of `RealPPPort.c` in `dist/Devel-PPPort`.  Relevant part of `make test_prep` output:
```
"../../miniperl" "-I../../lib" "../../lib/ExtUtils/xsubpp" -noprototypes -typemap '/usr/home/jkeenan/gitwork/perl/dist/Devel-PPPort/../../lib/ExtUtils/typemap' -typemap '/usr/home/jkeenan/gitwork/perl/dist/Devel-PPPort/typemap'  RealPPPort.xs > RealPPPort.xsc
mv RealPPPort.xsc RealPPPort.c
gcc9 -c    -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_FORTIFY_SOURCE=2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings -O2    -DVERSION=\"3.63\"  -DXS_VERSION=\"3.63\" -DPIC -fPIC "-I../.."  -W -Wall -Wdeclaration-after-statement RealPPPort.c
In file included from ../../perl.h:7145,
                 from RealPPPort.xs:31:
../../inline.h: In function 'Perl_av_store_simple':
../../inline.h:91:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   91 |     SV **ary = AvARRAY(av);
      |     ^~
gcc9 -c    -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_FORTIFY_SOURCE=2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings -O2    -DVERSION=\"3.63\"  -DXS_VERSION=\"3.63\" -DPIC -fPIC "-I../.."  -W -Wall -Wdeclaration-after-statement module2.c
In file included from ../../perl.h:7145,
                 from module2.c:17:
../../inline.h: In function 'Perl_av_store_simple':
../../inline.h:91:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   91 |     SV **ary = AvARRAY(av);
      |     ^~
gcc9 -c    -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_FORTIFY_SOURCE=2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings -O2    -DVERSION=\"3.63\"  -DXS_VERSION=\"3.63\" -DPIC -fPIC "-I../.."  -W -Wall -Wdeclaration-after-statement module3.c
In file included from ../../perl.h:7145,
                 from module3.c:17:
../../inline.h: In function 'Perl_av_store_simple':
../../inline.h:91:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   91 |     SV **ary = AvARRAY(av);
      |     ^~
rm -f ../../lib/auto/Devel/PPPort/PPPort.so
```
Relevant part of `inline.h`:
```
  81 PERL_STATIC_INLINE SV**
  82 Perl_av_store_simple(pTHX_ AV *av, SSize_t key, SV *val)
  83 {
  84     PERL_ARGS_ASSERT_AV_STORE_SIMPLE;
  85     assert(SvTYPE(av) == SVt_PVAV);
  86     assert(!SvMAGICAL(av));
  87     assert(!SvREADONLY(av));
  88     assert(AvREAL(av));
  89     assert(key > -1);
  90 
  91     SV **ary = AvARRAY(av);
  92 
```

**Comment by smpetershm** [comment]
@nwc10 Just watching the `-Wdeclaration-after-statement` that I added because GCC 3.0 stopped enforcing it and my OpenBSD 368 was stuck on 2.95.  Back in the old days when `-Wall` was really scary.

**Comment by smpetershm** [comment]
> Almost as good news: I tried this branch on FreeBSD-12 using `gcc9` (the default version of `gcc` on this platform). There were 3 new instances of the `-Wdeclaration-after-statement` warnings category. I suspect these warnings can be remediated fairly easily.

C99 allows for mixed declarations and statements so checking for `-Wdeclaration-after-statement` won't be necessary for this change or after.

**Comment by jkeenan** [comment]
On 10/6/21 2:27 PM, Steve Peters wrote:
>     Almost as good news: I tried this branch on FreeBSD-12 using |gcc9|
>     (the default version of |gcc| on this platform). There were 3 new
>     instances of the |-Wdeclaration-after-statement| warnings category.
>     I suspect these warnings can be remediated fairly easily.
> 
> C99 allows for mixed declarations and statements so checking for 
> |-Wdeclaration-after-statement| won't be necessary for this change or after.
> 

How then do we quiet this warning, at least with respect to these 3 
instances?

**Comment by jkeenan** [comment]
> On 10/6/21 2:27 PM, Steve Peters wrote: Almost as good news: I tried this branch on FreeBSD-12 using |gcc9| (the default version of |gcc| on this platform). There were 3 new instances of the |-Wdeclaration-after-statement| warnings category. I suspect these warnings can be remediated fairly easily. C99 allows for mixed declarations and statements so checking for |-Wdeclaration-after-statement| won't be necessary for this change or after.
> How then do we quiet this warning, at least with respect to these 3 instances?

At least two smoke-testing rigs are set up to build with `-Accflags="-Werror=declaration-after-statement"` and is failing as a result.  See:  http://perl.develop-help.com/raw/?id=273699 or http://perl.develop-help.com/raw/?id=273700.

**Comment by nwc10** [comment]
> How then do we quiet this warning, at least with respect to these 3 instances?

Commit 0081bff5fe6df5 should fix this.

There might be a better solution, but if so it's not obvious to me.

**Comment by nwc10** [comment]
> At least two smoke-testing rigs are set up to build with `-Accflags="-Werror=declaration-after-statement"` and is failing as a result.

commit 6b5b56c60ca7e claims its first two scalps.

When policy was strictly C89 on this, those flags were a really good idea - they have been keeping us "honest" and spotting whether we mess up on the C89-strictures. But with a move to permitting C99 mixed declarations and code, that means that (for blead, but not maint until maint-5.36) these configurations are now wrong. This branch renders that option obsolete for us.

**Comment by jkeenan** [comment]
> > At least two smoke-testing rigs are set up to build with `-Accflags="-Werror=declaration-after-statement"` and is failing as a result.
> 
> commit [6b5b56c](https://github.com/Perl/perl5/commit/6b5b56c60ca7e6dafca8e6265e6c6622f186dc7e) claims its first two scalps.
> 

Fix confirmed:
```
$ report-build-warnings 50a17201c9.freebsd.gcc9.threaded.smoke-me-nicholas-C99.maketp.output.txt.gz 
File:  50a17201c9.freebsd.gcc9.threaded.smoke-me-nicholas-C99.maketp.output.txt.gz

  Wc++-compat                              240
  Wempty-body                                6
  Wsign-compare                              4
  Wtype-limits                               2
```
> When policy was strictly C89 on this, those flags were a really good idea - they have been keeping us "honest" and spotting whether we mess up on the C89-strictures. But with a move to permitting C99 mixed declarations and code, that means that (for blead, but not maint until maint-5.36) these configurations are now wrong. This branch renders that option obsolete for us.

Should we now document that configuring with `-Accflags="-Werror=declaration-after-statement"` is now inappropriate?  (I want to have the correct info before I contact this smoke-tester.)

**Comment by jkeenan** [comment]
> > At least two smoke-testing rigs are set up to build with `-Accflags="-Werror=declaration-after-statement"` and is failing as a result.
> 
> commit [6b5b56c](https://github.com/Perl/perl5/commit/6b5b56c60ca7e6dafca8e6265e6c6622f186dc7e) claims its first two scalps.
> 

[Grepping CPAN](https://grep.metacpan.org/search?size=20&q=Werror%3Ddeclaration-after-statement&qd=&qft=) suggests that there may be other scalps to come along in Sereal-Decoder, -Encoder, -Merger, -Path, -Splitter and YAML-Safe.

**Comment by nwc10** [comment]
> Should we now document that configuring with `-Accflags="-Werror=declaration-after-statement"` is now inappropriate? (I want to have the correct info before I contact this smoke-tester.)

I guess that we should wait until reviewers have said that this PR is good to merge.

Also, I'm not sure **where** to document that this is no longer a good smoke test configuration **for blead** - does it belong in core, or is there somewhere in the smoker ecoystem?

**Comment by nwc10** [comment]
> [Grepping CPAN](https://grep.metacpan.org/search?size=20&q=Werror%3Ddeclaration-after-statement&qd=&qft=) suggests that there may be other scalps to come along in Sereal-Decoder, -Encoder, -Merger, -Path, -Splitter and YAML-Safe.

Thanks. I was aware of Sereal, hence Sereal/Sereal#265 (partly because in turn Yves is aware of this C99 work and the implications).

I wasn't aware of YAML-Safe, but I know that Cpanel::JSON::XS will need a tweak.

I was assuming that PRs (or similar) on these modules should wait until we're ready to merge, and was hoping that there would be a week or two between that point and when 5.35.5 ships, such that we have time for everyone to be ready.

**Comment by jkeenan** [comment]
> > Should we now document that configuring with `-Accflags="-Werror=declaration-after-statement"` is now inappropriate? (I want to have the correct info before I contact this smoke-tester.)
> 
> I guess that we should wait until reviewers have said that this PR is good to merge.
> 

Okay.

> Also, I'm not sure **where** to document that this is no longer a good smoke test configuration **for blead** - does it belong in core, or is there somewhere in the smoker ecoystem?

Presumably we would want to discourage **anyone** -- smoke-tester or otherwise -- from configuring with `-Accflags="-Werror=declaration-after-statement"` going forward.  That suggests that the documentation should be in `INSTALL`, perhaps under *Major Configure-time Build Options*.

File: pod/perlhacktips.pod
Comment (tonycoz) [inline pod/perlhacktips.pod:200]: Accessing beyond message[0] is technically undefined behaviour, so we avoid that.  I don't know of anywhere it's a practical issue though.

File: pod/perlhacktips.pod
Comment (tonycoz) [inline pod/perlhacktips.pod]: Perl's own formatters support %zu, so this might be confusing.

File: README.win32
Comment (nwc10) [reply README.win32]: Well spotted. Fixed. Thanks.

File: README.win32
Comment (nwc10) [reply README.win32]: I removed IA64 entirely, and moved the mention of x86_64 to the earlier paragraph shown.

And then realised that it makes mingw a bit ambiguous without further clarification - unlike the others mingw is just a 32 bit toolchain, but what it build can **run** on 64 bit Windows. So I tweaked the text further as part of 0788bfbebf59728

I assume that if we like this (or we like it with further edits) then I will merge that commit/all commits changing README.win32 into one commit will all the changes, and a unified commit message.

File: README.win32
Comment (nwc10) [reply README.win32]: Done as part of commit 0788bfbebf59728 - see previous comment about merge/rebase.

File: pod/perlhacktips.pod
Comment (nwc10) [reply pod/perlhacktips.pod]: Yes, well spotted. `Perl_sv_vcatpvfn_flags` takes `%`. It's only the C library `*printf` functions where this matters, and we "outlaw" most of those except `sprintf`. I was typing "maybe I should scrub this" but I think it better to clarify `sprintf` vs sv_catpvf etc, as we do use `sprintf` in our C code, and it will be a trap.

Comment (nwc10) [reply pod/perlhacktips.pod]: I tried to clarify this in commit 41a65e5d9d44cfe10f3 - is this better?

**Comment by jkeenan** [comment]
> > > Should we now document that configuring with `-Accflags="-Werror=declaration-after-statement"` is now inappropriate? (I want to have the correct info before I contact this smoke-tester.)
> > 
> > 
> > I guess that we should wait until reviewers have said that this PR is good to merge.
> 
> Okay.
> 
> > Also, I'm not sure **where** to document that this is no longer a good smoke test configuration **for blead** - does it belong in core, or is there somewhere in the smoker ecoystem?
> 
> Presumably we would want to discourage **anyone** -- smoke-tester or otherwise -- from configuring with `-Accflags="-Werror=declaration-after-statement"` going forward. That suggests that the documentation should be in `INSTALL`, perhaps under _Major Configure-time Build Options_.

Some more data which, IMO, supports the notion of explicitly documenting that, going forward, configuring with `-Accflags="-Werror=declaration-after-statement"` is a Bad Thing.

After @nwc10 fixed the parts of Devel-PPPort that were triggering 3 instances of this category of warnings, I decided to see what would happen when I *did* configure with those ccflags and compiled with `gcc`.  How far would I get?

I got through `make`, so `perl` was built.  However, during the test suite I experienced these test failures:
```
$ cd t;TEST_JOBS=1 ./perl harness ../dist/ExtUtils-ParseXS/t/*.t; cd -;extract-sha;gitcurr
../dist/ExtUtils-ParseXS/t/001-basic.t ............................ 3/18 In file included from /home/jkeenan/gitwork/perl/perl.h:7145,
                 from XSTest.xs:2:
/home/jkeenan/gitwork/perl/inline.h: In function ‘Perl_av_store_simple’:
/home/jkeenan/gitwork/perl/inline.h:91:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
   91 |     SV **ary = AvARRAY(av);
      |     ^~
cc1: some warnings being treated as errors
error building XSTest.o from 'XSTest.c' at /home/jkeenan/gitwork/perl/dist/ExtUtils-ParseXS/../../lib/ExtUtils/CBuilder/Base.pm line 185.
# Looks like your test exited with 2 just after 3.
../dist/ExtUtils-ParseXS/t/001-basic.t ............................ Dubious, test returned 2 (wstat 512, 0x200)
Failed 15/18 subtests 
../dist/ExtUtils-ParseXS/t/002-more.t ............................. 1/32 In file included from /home/jkeenan/gitwork/perl/perl.h:7145,
                 from XSMore.xs:2:
/home/jkeenan/gitwork/perl/inline.h: In function ‘Perl_av_store_simple’:
/home/jkeenan/gitwork/perl/inline.h:91:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
   91 |     SV **ary = AvARRAY(av);
      |     ^~
cc1: some warnings being treated as errors
error building XSMore.o from 'XSMore.c' at /home/jkeenan/gitwork/perl/dist/ExtUtils-ParseXS/../../lib/ExtUtils/CBuilder/Base.pm line 185.
# Looks like your test exited with 2 just after 2.
../dist/ExtUtils-ParseXS/t/002-more.t ............................. Dubious, test returned 2 (wstat 512, 0x200)
Failed 30/32 subtests 
../dist/ExtUtils-ParseXS/t/003-usage.t ............................ 1/24 In file included from /home/jkeenan/gitwork/perl/perl.h:7145,
                 from XSUsage.xs:2:
/home/jkeenan/gitwork/perl/inline.h: In function ‘Perl_av_store_simple’:
/home/jkeenan/gitwork/perl/inline.h:91:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
   91 |     SV **ary = AvARRAY(av);
      |     ^~
cc1: some warnings being treated as errors
error building XSUsage.o from 'XSUsage.c' at /home/jkeenan/gitwork/perl/dist/ExtUtils-ParseXS/../../lib/ExtUtils/CBuilder/Base.pm line 185.
# Looks like your test exited with 2 just after 2.
../dist/ExtUtils-ParseXS/t/003-usage.t ............................ Dubious, test returned 2 (wstat 512, 0x200)
Failed 22/24 subtests 
../dist/ExtUtils-ParseXS/t/101-standard_typemap_locations.t ....... ok
[snip]
../dist/ExtUtils-ParseXS/t/600-t-compat.t ......................... ok   

Test Summary Report
-------------------
../dist/ExtUtils-ParseXS/t/001-basic.t                          (Wstat: 512 Tests: 3 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 18 tests but ran 3.
../dist/ExtUtils-ParseXS/t/002-more.t                           (Wstat: 512 Tests: 2 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 32 tests but ran 2.
../dist/ExtUtils-ParseXS/t/003-usage.t                          (Wstat: 512 Tests: 2 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 24 tests but ran 2.
Files=27, Tests=201,  4 wallclock secs ( 0.06 usr  0.02 sys +  3.23 cusr  0.30 csys =  3.61 CPU)
Result: FAIL

SHA:    50a17201c9
Branch: smoke-me/nicholas/C99

```
The same ParseXS test failures can be observed in more recent smoke-testing reports from the two rigs upon which I reported yesterday.  See:  http://perl.develop-help.com/raw/?id=273752 and http://perl.develop-help.com/raw/?id=273751.  These rigs were originally failing during `make`.  They're getting past that point now, but failing in ParseXS.

While I think it's practical for us correct the ParseXS errors and most of the points (Encode, Cpanel-JSON-XS) on CPAN that we discussed yesterday, it will not be practical for us to errors beyond that.  Hence, the need for explicitly recommending against `-Accflags="-Werror=declaration-after-statement"`.

File: pod/perlhacktips.pod
Comment (tonycoz) [reply pod/perlhacktips.pod]: That looks good.

**Comment by craigberry** [comment]
The branch builds fine with:

```
$ cc/vers
VSI C V7.4-002 on OpenVMS Alpha V8.4-2L1
```

It doesn't show up in the compiler version, but this particular system does _not_ have the library/header enhancements added in 2020 to complete C99 support.

**Review by craigberry — APPROVED** [review]

**Comment by nwc10** [comment]
> After @nwc10 fixed the parts of Devel-PPPort that were triggering 3 instances of this category of warnings, I decided to see what would happen when I _did_ configure with those ccflags and compiled with `gcc`. How far would I get?

I re-read this and thought *what the heck*?

Thanks for testing this. I had simply assumed that it would not work.

However, it did. Because...

The probe in `Configure` was for `long long` and declarations in a for loop. There was no declaration after a statement in it, hence adding these flags didn't cause the probe to fail and `Configure` abort.

*Then* we have `cflags.SH` which attempts to add more warning flags for `gcc`. These flags originate in the time of C89, and at least on some gcc versions `-Wextra -W` ended up turning on `-Wlong-long -Wdeclaration-after-statement`, hence `cflags.SH` adding *no* variants of those to get back to zero.

Ultimately the C flags looked like this:

```
cc -c -DPERL_CORE -Werror=declaration-after-statement -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -std=c99 -O2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings mathoms.c
```

and so that `-Wno-declaration-after-statement` disables the error. And with it last the core builds.

**Mostly**

What we actually have is

```
ccflags='-Werror=declaration-after-statement -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
ccwarnflags=' -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings'
```

and the core (and ExtUtils::MakeMaker by default) uses both the "flags" and the "warnings flags". Where what we have here is the nonsense that the "warnings flags" actually unbreak things.

The ExtUtils::ParseXS tests don't use the warnings flags and as a consequence the regular flags alone don't work.

Thanks for testing things, and finding that we had a problem that I had completely missed.

> While I think it's practical for us correct the ParseXS errors and most of the points (Encode, Cpanel-JSON-XS) on CPAN that we discussed yesterday, it will not be practical for us to errors beyond that. Hence, the need for explicitly recommending against `-Accflags="-Werror=declaration-after-statement"`.

So.

In the rebased branch

* The probe in `Configure` also tests for declarations after statements
* Hence attempting `-Accflags="-Wdeclaration-after-statement"` is a configuration fail. You can't even generate `config.sh` without manual intervention
* `cflags.SH` has another consistency fix - it no longer adds `-std=c99` if there is already `-std=gnu99` in the `cflags`
* I added a section to `INSTALL` mentioning the C99 changes


I've tested this again on AIX, HP/UX and the ancient gcc on dromedary, and all seems good. I guess give it another 24 hours or so to find any other smoke-me failures, and then we're good to merge.

~~@jkeenan - could you contact the owners of the smokers that are configured to add `-Accflags="-Wdeclaration-after-statement"` and ask them to remove it?~~ -- I contacted George Greer - is anyone else smoking with these flags?

File: README.win32
Comment (nwc10) [reply README.win32]: @steve-m-hay I think that the text now in README.win32 fixes the problem you spotted, so I'm marking this "conversation" as resolved.

If I got this wrong, please accept my apologies and re-open this "conversation".

**Comment by nwc10** [comment]
I believe that all Sereal modules on CPAN share the same repository. Dan Kogai has already released an updated Encode. He's fast!

PRs to distributions that add `-Werror-=declaration-after-statement` unconditionally

* Sereal/Sereal#265
* rurban/Cpanel-JSON-XS#186
* rurban/YAML-Safe#3

And to send our `Makefile.PL` change upstream, Dual-Life/Devel-PPPort#211

**Pull request closed without merging**
